web-ui/src/main/resources/catalog/components/viewer/projectionSwitcher/ProjectionSwitcher.js

```diff
@@ -0,0 +1,175 @@
+   *              that allows you to change the projection.
+   */
+  module.directive('gnProjectionSwitcher', [
+    'gnViewerSettings',
+    function(gnViewerSettings) {
+      return {
+        restrict: 'A',
+        templateUrl: '../../catalog/components/viewer/projectionSwitcher/' +
+            'partials/projectionSwitcher.html',
+        scope: {
+          map: '=gnProjectionSwitcher'
+        },
+        controllerAs: 'gnProjectionSwitcherCtrl',
+        controller: ['$scope', 'gnViewerSettings', 'gnMap', '$rootScope', 
+          function(scope, gnViewerSettings, gnMap, $rootScope) {
+
+          scope.changeLayerProjection = function(
+            layer, oldProj, newProj) {
+            if (layer instanceof ol.layer.Group) {
+              layer.getLayers().forEach(
+                  function(subLayer) {
+                    this.changeLayerProjection(
+                      subLayer, oldProj,
+                      newProj);
+                  });
+            } else if (layer instanceof ol.layer.Tile) {
+              var tileLoadFunc = layer.getSource()
+                .getTileLoadFunction();
+              layer.getSource().setTileLoadFunction(
+                tileLoadFunc);
+            } else if (layer instanceof ol.layer.Vector) {
+              var features = layer.getSource().getFeatures();
+              for (var i = 0; i < features.length; i += 1) {
+                features[i].getGeometry().transform(
+                  oldProj, newProj);
+              }
+            }
+          };
+          
+          scope.switchProjection = function(projection) {
+              var view = scope.map.getView();
+              var oldProj = view.getProjection();
+              var newProj = ol.proj.get(projection);
+              
+              if(oldProj == newProj) {
+                return;
+              }
+              
+              var projectionConfig = {};
+              
+              $.each(gnViewerSettings.mapConfig.switcherProjectionList, function(i, config) {
+                if(config['code'] == projection) {
+                  projectionConfig = config;
+                }
+              });
+              
+              var newExtent = ol.proj.transformExtent(
+                  view.calculateExtent(scope.map.getSize()), oldProj, newProj);
+              
+              if(newProj.getExtent()) {
+                newExtent = newProj.getExtent();
+              }
+              
+              var newCenter = ol.proj.transform(scope.map.getView().getCenter(), oldProj, newProj);
+              
+              if(projectionConfig.center) {
+                newCenter = JSON.parse(projectionConfig.center);
+              }
+
+              var mapsConfig = {
+                projection: newProj,
+                extent: newExtent,
+                center: newCenter
+              };
+              
+              if (projectionConfig.resolutions) {
```

Comment:
Again, I don't think is really needed here. Unless some people specifically requested that when a projection is used, the view should be constrained in zoom levels?

Reply:
Same as before.